### ISSUE TITLE
3.142.2 can't find Firefox on FreeBSD

### ISSUE THREAD

**Issue by PeterMozesMerl** [SeleniumHQ/selenium#7219]
## 🐛 Bug Report

Upgrading to 3.142.2 made selenium-webdriver lose its interest in running my examples and depress me with throwing "`page.current_path` raised exception: #Selenium::WebDriver::Error::SessionNotCreatedError: Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line" exceptions.

## To Reproduce

1. Upgrade selenium-webdriver from 3.141.0 to 3.142.2
2. Try running your examples

## Expected behavior

Selenium finds firefox.

## Environment

FreeBSD 11.2-RELEASE-p10, more or less default install on multiple servers
Normal firefox install (pkg install firefox): firefox-66.0.5,1
Rails 5.2.3
Capybara 3.20.1

I don’t have firefox’ path set in my suite because I had never needed it to be set. I had no similar issue before 3.142.2.

3.142.2 works on MacOS Mojave
3.141.0 works on my FreeBSD CI servers (as well as on MacOS)
3.142.2 doesn’t work on FreeBSD.

The firefox binary (/usr/local/bin/firefox) is in the path.

**Comment by p0deje** [comment]
Interesting. Assuming Firefox is in the page - it must be found by GeckoDriver.

1. What GeckoDriver version do you use?
2. Can you provide full stacktrace of error that you see?

**Comment by PeterMozesMerl** [comment]
1. It’s the latest. 0.24.0

2. The 98th line of capybara_general_helper.rb is ```find css, options```

```
  28) User / Profile / Actions Adding and removing user to groups
      Got 0 failures and 2 other errors:

      28.1) Failure/Error: Capybara.current_session.current_window.resize_to 1500, CAPYBARA_HOST_IS_DEVELOPMENT_MACHINE ? 1200 : 1500

            Selenium::WebDriver::Error::SessionNotCreatedError:
              Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:72:in `assert_ok'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:34:in `initialize'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `create_response'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/default.rb:114:in `request'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:64:in `call'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:167:in `execute'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:102:in `create_session'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/marionette/driver.rb:44:in `initialize'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/common/driver.rb:54:in `for'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver.rb:88:in `for'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/selenium/driver.rb:32:in `browser'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/selenium/driver.rb:139:in `current_window_handle'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/session.rb:438:in `current_window'
            # ./spec/support/includes/configuration/rspec_callback_helper.rb:34:in `block (2 levels) in <main>'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:123:in `block in run'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `loop'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `run'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:37:in `block (2 levels) in setup'

      28.2) Failure/Error: find css, **options.merge({visible: visible})

            Selenium::WebDriver::Error::SessionNotCreatedError:
              Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:72:in `assert_ok'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:34:in `initialize'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `create_response'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/default.rb:114:in `request'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:64:in `call'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:167:in `execute'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:102:in `create_session'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/marionette/driver.rb:44:in `initialize'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/common/driver.rb:54:in `for'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver.rb:88:in `for'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/selenium/driver.rb:32:in `browser'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/selenium/driver.rb:346:in `find_context'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/selenium/extensions/find.rb:17:in `find_by'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/selenium/extensions/find.rb:11:in `find_css'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/node/base.rb:107:in `find_css'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/queries/selector_query.rb:190:in `find_nodes_by_selector_format'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/queries/selector_query.rb:125:in `block in resolve_for'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/node/base.rb:77:in `synchronize'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/queries/selector_query.rb:124:in `resolve_for'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/node/finders.rb:295:in `block in synced_resolve'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/node/base.rb:83:in `synchronize'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/node/finders.rb:293:in `synced_resolve'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/node/finders.rb:48:in `find'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/session.rb:744:in `block (2 levels) in <class:Session>'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.14.0/lib/capybara/dsl.rb:51:in `block (2 levels) in <module:DSL>'
            # ./spec/old/support/capybara/capybara_general_helper.rb:98:in `wait_for_element'
            # ./spec/old/support/capybara/capybara_general_helper.rb:108:in `wait_for_render'
            # ./spec/support/includes/configuration/rspec_callback_helper.rb:64:in `block (2 levels) in <main>'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:123:in `block in run'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `loop'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `run'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
```

**Comment by twalpole** [comment]
The stack trace shows that you're running capybara 3.14 not the 3.20.1 you stated in your original report.  What other gems are changing between your swap from seleniuv-webdriver 3.141.0 to 3.142.2?

It's slightly confusing that your first command for the session is a `find`  without having actually visited any url - unless possibly you're using webmock and it's incorrectly mocked some of the selenium/capybara communications?? What does your driver registration look like? What are `wait_for_render` and `wait_for_element` doing?

**Comment by PeterMozesMerl** [comment]
There was a note at the end of my comment that this trace was with Capybara 3.14.0. When I see hundreds of examples fail after a bundle update, I try to revert the gems one by one to find the source of the error. The trace looks the same with Capybara 3.14 and 3.20. Right now all my examples pass with having ```gem 'selenium-webdriver', '=3.141.0'``` in the Gemfile.

What you see are two exceptions. None of them are in the example.

The first exception raised here:
```ruby
	config.before :each, js: true do
		Capybara.current_session.current_window.resize_to 1500, CAPYBARA_HOST_IS_DEVELOPMENT_MACHINE ? 1200 : 1500
	end
```

The second one is from here:

```ruby
	config.after :each, js: true do |example|
[...]
			wait_for_render
			visit 'transparent.png'
		end
	end
```

The purpose of the after callback is to eliminate (minimize) the deadlocks in the PostgreSQL database that happen when rogue Ajax calls are executed when the example is already finished and the DatabaseCleaner gem is deleting the table. I’m using delete instead of truncate after the features which makes them happen ten times less often, but more often than optimal (that would be zero). My CI servers use parallel_tests gem. It runs the tests in a separated database per process hence it’s not the direct reason of the deadlocks. It is the fact that 8-12 Firefox processes run in parallel. Under heavy load, the rogue Ajax calls happen more often.

The retry is there for a related reason. It’s only set for Net::ReadTimeout that also happens once in a while under heavy load.

```ruby

	# Only retry when Selenium raises Net::ReadTimeout
	config.exceptions_to_retry = [Net::ReadTimeout]

	# Each feature should wait for the page to finish loading (and pending ajax calls) except for the Net::ReadTimeout
	config.after :each, js: true do |example|
		rspec_log 'After each js: true'
		# Neither wait nor save screenshot on Net::ReadTimeout
		if example.exception.present? && example.exception.class == Net::ReadTimeout
			rspec_log 'Won’t wait as a Net::ReadTimeout raised', error: true
		else
			# Save the screenshot if there was an exception
			screenshot_and_save_page if example.exception.present?
			# Navigate away from the page to avoid deadlocks and timeouts caused by ajax calls
			wait_for_render
			visit 'transparent.png'
		end
	end

	# Wait for onLoad
	def wait_for_load_done
		wait_for_element 'body.load-done'
	end

	# Wait until the page has been rendered
	def wait_for_render
		wait_for_element '#footer', visible: false
		wait_for_ajax
		wait_for_load_done
	end

	# Wait for ajax
	def wait_for_ajax
		capybara_loop { evaluate_script("typeof jQuery === 'function' && jQuery.active === 0 && document.readyState === 'complete'").present? }
	end
```

Normally, find is never called before visiting a page because the ```Capybara.current_session.current_window.resize_to``` is not supposed to throw an Exception. If I comment it out, then something else will, because Firefox is not found. I tried it.

**Comment by twalpole** [comment]
Assuming you setup your after chain correctly (database_cleaner is run in an `append_after` block rather than an `after` block) so the database isn't cleaned until after Capybaras reset is called there should be no way for Ajax calls to happen while/after the database is being cleared since Capybaras reset waits for all requests to complete.  Getting that correct would fix some of the complexity you appear to have needlessly built into your setup - but it probably won't fix the current issue. If you could try the other released versions of selenium-webdriver (3.141.59, 3.141.592, 3.141.5926, 3.142.0, 3.142.1 ) it could help to narrow down the cause of your current issue.

**Comment by PeterMozesMerl** [comment]
So my application is in the 5% of the cases, I guess. 

I have no idea why you believe you have a better understanding of my application without seeing it at all than I have after developing it for many years and serving millions of active users. I understand that blog or webshop applications use a single database, without replication, distribution, and intentional begin...rollback blocks. Good luck for testing these with the built-in transactional fixtures.

When I report a bug, I don’t do it to criticize your work. I consider it helping. I also consider it normal because I benefit from your work. Others might think the same as I do because it’s the first time in my life that after a bug report, a developer is trying to "fix" my software instead of the bug.

If the price of fixing it on FreeBSD is that I make some unneeded noise in my CI servers’ logs, then I’m willing to pay it by committing a version that has no resize to before hook and will throw the same exception on something else. As I wrote to you, I don’t have the issue on my desktop, which is MacOS.

**Comment by PeterMozesMerl** [comment]
```
34) Session / Sign in Wrong login Wrong username
      Got 0 failures and 2 other errors:

      34.1) Failure/Error: visit root_path

            Selenium::WebDriver::Error::SessionNotCreatedError:
              Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:72:in `assert_ok'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:34:in `initialize'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `create_response'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/default.rb:114:in `request'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:64:in `call'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:167:in `execute'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:102:in `create_session'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/marionette/driver.rb:44:in `initialize'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/common/driver.rb:54:in `for'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver.rb:88:in `for'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.20.1/lib/capybara/selenium/driver.rb:48:in `browser'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.20.1/lib/capybara/selenium/driver.rb:67:in `visit'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.20.1/lib/capybara/session.rb:277:in `visit'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/capybara-3.20.1/lib/capybara/dsl.rb:51:in `block (2 levels) in <module:DSL>'
            # ./spec/old/features/session/session_sign_in_spec.rb:321:in `block (3 levels) in <top (required)>'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:123:in `block in run'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `loop'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:110:in `run'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
            # /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/rspec-retry-0.6.1/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
```

The related part in session_sign_in_spec.rb

```
		scenario 'Wrong username' do
			visit root_path
```

I hope the way this exception was raised was canonical enough for you to focus on fixing selenium-webdriver on FreeBSD.

I commented out the before :each callback, and reverted to the previous selenium-webdriver gem.

Here is the full diff of the Gemfile.lock. I can also ask my doctor what my blood type is if you need it to fix the bug.

```
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -88,8 +88,8 @@ GEM
     capybara-screenshot (1.0.22)
       capybara (>= 1.0, < 4)
       launchy
-    childprocess (0.9.0)
-      ffi (~> 1.0, >= 1.0.11)
+    childprocess (1.0.1)
+      rake (< 13.0)
     chunky_png (1.3.11)
     ci_reporter (2.0.0)
       builder (>= 2.1.2)
@@ -333,8 +333,8 @@ GEM
       tilt (>= 1.1, < 3)
     select2-rails (4.0.3)
       thor (~> 0.14)
-    selenium-webdriver (3.141.0)
-      childprocess (~> 0.5)
+    selenium-webdriver (3.142.2)
+      childprocess (>= 0.5, < 2.0)
       rubyzip (~> 1.2, >= 1.2.2)
     sexp_processor (4.12.0)
     shoulda-matchers (4.0.1)
@@ -456,7 +456,7 @@ DEPENDENCIES
   rubycritic
   sass-rails (~> 5.0)
   select2-rails
-  selenium-webdriver (= 3.141.0)
+  selenium-webdriver
   shoulda-matchers
   simplecov
   slim-rails
```

Note: all my features pass on FreeBSD with ```gem 'selenium-webdriver', '=3.141.0'``` (and I’ve never had this issue before). All fail with the current version. There was no other difference than the gem version.

**Comment by p0deje** [comment]
So the issue is definitely coming from GeckoDriver response, but it seems like Selenium might send malformed payload to it. Can you please add the following line somewhere before the browser is initialized and paste results for 3.141.0 and 3.142.2:

```ruby
Selenium::WebDriver.logger.level = :debug
```

Is it possible for you to just run Selenium without your testing stack? It might be easier if we know this is exactly coming from Selenium and not other parts.

Finally, do you have WebDrivers gem installed?

**Comment by PeterMozesMerl** [comment]
I have no WebDrivers installed. I ran the code you wrote above with 3.141.0 as well as 3.142.2 under Rails.

```
$ gem list | grep -i driver
chromedriver-helper (2.1.1, 2.1.0)
selenium-webdriver (3.142.2, 3.141.0)
websocket-driver (0.7.0)


$ mcedit Gemfile
$ bundle
Fetching gem metadata from https://rubygems.org/.........
[...]
Using selenium-webdriver 3.141.0 (was 3.142.2)
[...]

$ RAILS_ENV=test rails runner selenium.rb
2019-05-20 10:00:02 DEBUG Selenium Executing Process ["/usr/local/jenkins/bin/geckodriver", "--binary=/usr/local/bin/firefox", "--port=4444"]
2019-05-20 10:00:02 DEBUG Selenium polling for socket on ["127.0.0.1", 4444]
2019-05-20 10:00:02 INFO Selenium Skipping handshake as we know it is W3C.
2019-05-20 10:00:02 INFO Selenium -> POST session
2019-05-20 10:00:02 INFO Selenium    >>> http://127.0.0.1:4444/session | {"desiredCapabilities":{"browserName":"firefox","marionette":true,"timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}},"capabilities":{"firstMatch":[{"browserName":"firefox","timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}}]}}
2019-05-20 10:00:02 DEBUG Selenium      > {"Accept"=>"application/json", "Content-Type"=>"application/json; charset=UTF-8", "User-Agent"=>"selenium/3.141.0 (ruby unix)", "Content-Length"=>"247"}
1558339202422	mozrunner::runner	INFO	Running command: "/usr/local/bin/firefox" "-marionette" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofile.jIKmA7mEl60q"
1558339202423	geckodriver::marionette	DEBUG	Waiting 60s to connect to browser on 127.0.0.1:48348
Error: no DISPLAY environment variable specified
1558339202527	mozrunner::runner	DEBUG	Killing process 48169
1558339202527	webdriver::server	DEBUG	<- 500 Internal Server Error {"value":{"error":"unknown error","message":"invalid argument: can't kill an exited process","stacktrace":""}}
2019-05-20 10:00:02 INFO Selenium <- {"value":{"error":"unknown error","message":"invalid argument: can't kill an exited process","stacktrace":""}}
Traceback (most recent call last):
	31: from bin/rails:4:in `<main>'
	30: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
	29: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
	28: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `block in require'
	27: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	26: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require_with_bootsnap_lfi'
	25: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
	24: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `block in require_with_bootsnap_lfi'
	23: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require'
	22: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/commands.rb:18:in `<main>'
	21: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/command.rb:46:in `invoke'
	20: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/command/base.rb:65:in `perform'
	19: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	18: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	17: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	16: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/commands/runner/runner_command.rb:38:in `perform'
	15: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:59:in `load'
	14: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:59:in `load'
	13: from /usr/local/jenkins/workspace/xxx/selenium.rb:6:in `<main>'
	12: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver.rb:86:in `for'
	11: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/common/driver.rb:52:in `for'
	10: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/firefox/driver.rb:31:in `new'
	 9: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/firefox/driver.rb:31:in `new'
	 8: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/firefox/marionette/driver.rb:50:in `initialize'
	 7: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/bridge.rb:99:in `create_session'
	 6: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/bridge.rb:166:in `execute'
	 5: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/http/common.rb:62:in `call'
	 4: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/http/default.rb:104:in `request'
	 3: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/http/common.rb:84:in `create_response'
	 2: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/http/common.rb:84:in `new'
	 1: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/response.rb:32:in `initialize'
/usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/response.rb:69:in `assert_ok': invalid argument: can't kill an exited process (Selenium::WebDriver::Error::UnknownError)

$ mcedit Gemfile
$ bundle update
Fetching gem metadata from https://rubygems.org/.........
[...]
Using selenium-webdriver 3.142.2 (was 3.141.0)
[...]

$ RAILS_ENV=test rails runner selenium.rb
2019-05-20 10:01:44 DEBUG Selenium Executing Process ["/usr/local/jenkins/bin/geckodriver", "--port=4444"]
2019-05-20 10:01:44 DEBUG Selenium polling for socket on ["127.0.0.1", 4444]
2019-05-20 10:01:45 INFO Selenium Skipping handshake as we know it is W3C.
2019-05-20 10:01:45 INFO Selenium -> POST session
2019-05-20 10:01:45 INFO Selenium    >>> http://127.0.0.1:4444/session | {"desiredCapabilities":{"browserName":"firefox","version":"","platform":"ANY","javascriptEnabled":false,"cssSelectorsEnabled":false,"takesScreenshot":false,"nativeEvents":false,"rotatable":false,"marionette":true,"timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}},"capabilities":{"firstMatch":[{"browserName":"firefox","timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}}]}}
2019-05-20 10:01:45 DEBUG Selenium      > {"Accept"=>"application/json", "Content-Type"=>"application/json; charset=UTF-8", "User-Agent"=>"selenium/3.142.2 (ruby unix)", "Content-Length"=>"394"}
1558339305050	webdriver::server	DEBUG	<- 500 Internal Server Error {"value":{"error":"session not created","message":"Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line","stacktrace":""}}
2019-05-20 10:01:45 INFO Selenium <- {"value":{"error":"session not created","message":"Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line","stacktrace":""}}
Traceback (most recent call last):
	31: from bin/rails:4:in `<main>'
	30: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
	29: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
	28: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `block in require'
	27: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	26: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require_with_bootsnap_lfi'
	25: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
	24: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `block in require_with_bootsnap_lfi'
	23: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require'
	22: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/commands.rb:18:in `<main>'
	21: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/command.rb:46:in `invoke'
	20: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/command/base.rb:65:in `perform'
	19: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	18: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	17: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	16: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/commands/runner/runner_command.rb:38:in `perform'
	15: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:59:in `load'
	14: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:59:in `load'
	13: from /usr/local/jenkins/workspace/xxx/selenium.rb:6:in `<main>'
	12: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver.rb:88:in `for'
	11: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/common/driver.rb:54:in `for'
	10: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
	 9: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
	 8: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/marionette/driver.rb:44:in `initialize'
	 7: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:102:in `create_session'
	 6: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:167:in `execute'
	 5: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:64:in `call'
	 4: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/default.rb:114:in `request'
	 3: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `create_response'
	 2: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `new'
	 1: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:34:in `initialize'
/usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:72:in `assert_ok': Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line (Selenium::WebDriver::Error::SessionNotCreatedError)
```

**Comment by PeterMozesMerl** [comment]
```
$ geckodriver --version
geckodriver 0.24.0

$ ls -l ~/bin/geckodriver
-rwxr-xr-x  1 jenkins  jenkins  10490621 Feb 16 16:30 /usr/local/jenkins/bin/geckodriver
```

The geckodriver is compiled with Rust. Should I recompile it?

I mean, although it’s the latest version of the source code, maybe some dependency has changed since February. I will give it a try.

**Comment by PeterMozesMerl** [comment]
I don’t know the Selenium internals at all. What I see in the dumps is that the desiredCapabilities differ.

```
# tcpdump dst port 4444 -A -i lo0 -v
tcpdump: listening on lo0, link-type NULL (BSD loopback), capture size 262144 bytes
11:56:05.213787 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60, bad cksum 0 (->3cba)!)
    localhost.59133 > localhost.krb524: Flags [S], cksum 0xfe30 (incorrect -> 0x116f), seq 2881554873, win 65535, options [mss 16344,nop,wscale 6,sackOK,TS val 126830856 ecr 0], length 0
E..<..@.@..............\..	..........0....?........
..I.....
11:56:05.480282 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60, bad cksum 0 (->3cba)!)
    localhost.59134 > localhost.krb524: Flags [S], cksum 0xfe30 (incorrect -> 0x31bf), seq 142904218, win 65535, options [mss 16344,nop,wscale 6,sackOK,TS val 126831123 ecr 0], length 0
E..<..@.@..............\.............0....?........
..J.....
11:56:05.480322 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59134 > localhost.krb524: Flags [.], cksum 0xfe28 (incorrect -> 0x9d89), ack 3337522481, win 1276, options [nop,nop,TS val 126831123 ecr 2386564536], length 0
E..4..@.@..............\.......1.....(.....
..J..@..
11:56:05.480359 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59134 > localhost.krb524: Flags [F.], cksum 0xfe28 (incorrect -> 0x9d88), seq 0, ack 1, win 1276, options [nop,nop,TS val 126831123 ecr 2386564536], length 0
E..4..@.@..............\.......1.....(.....
..J..@..
11:56:05.480687 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59134 > localhost.krb524: Flags [.], cksum 0xfe28 (incorrect -> 0x9d87), ack 2, win 1276, options [nop,nop,TS val 126831123 ecr 2386564536], length 0
E..4..@.@..............\.......2.....(.....
..J..@..
11:56:05.481776 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60, bad cksum 0 (->3cba)!)
    localhost.59135 > localhost.krb524: Flags [S], cksum 0xfe30 (incorrect -> 0x1650), seq 1830044279, win 65535, options [mss 16344,nop,wscale 6,sackOK,TS val 126831124 ecr 0], length 0
E..<..@.@..............\m.Bw.........0....?........
..J.....
11:56:05.481840 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59135 > localhost.krb524: Flags [.], cksum 0xfe28 (incorrect -> 0x31bb), ack 3108511056, win 1276, options [nop,nop,TS val 126831124 ecr 982464336], length 0
E..4..@.@..............\m.Bx.H.P.....(.....
..J.:.7P
11:56:05.482135 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 313, bad cksum 0 (->3bbd)!)
    localhost.59135 > localhost.krb524: Flags [P.], cksum 0xff2d (incorrect -> 0x051a), seq 0:261, ack 1, win 1276, options [nop,nop,TS val 126831124 ecr 982464336], length 261
E..9..@.@..............\m.Bx.H.P.....-.....
..J.:.7PPOST /session HTTP/1.1
Accept: application/json
Content-Type: application/json; charset=UTF-8
User-Agent: selenium/3.141.0 (ruby unix)
Content-Length: 247
Accept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3
Connection: close
Host: 127.0.0.1:4444


11:56:05.482161 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 299, bad cksum 0 (->3bcb)!)
    localhost.59135 > localhost.krb524: Flags [P.], cksum 0xff1f (incorrect -> 0xe60d), seq 261:508, ack 1, win 1276, options [nop,nop,TS val 126831124 ecr 982464336], length 247
E..+..@.@..............\m.C}.H.P...........
..J.:.7P{"desiredCapabilities":{"browserName":"firefox","marionette":true,"timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}},"capabilities":{"firstMatch":[{"browserName":"firefox","timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}}]}}
11:56:05.596098 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59135 > localhost.krb524: Flags [.], cksum 0xfe28 (incorrect -> 0x2dc6), ack 280, win 1272, options [nop,nop,TS val 126831239 ecr 982464451], length 0
E..4..@.@..............\m.Dt.H.g.....(.....
..J.:.7.
11:56:05.596490 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59135 > localhost.krb524: Flags [F.], cksum 0xfe28 (incorrect -> 0x2dc1), seq 508, ack 280, win 1276, options [nop,nop,TS val 126831239 ecr 982464451], length 0
E..4..@.@..............\m.Dt.H.g.....(.....
..J.:.7.
^C
11 packets captured
237 packets received by filter
0 packets dropped by kernel



# tcpdump dst port 4444 -A -i lo0 -v
tcpdump: listening on lo0, link-type NULL (BSD loopback), capture size 262144 bytes
11:55:28.099441 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60, bad cksum 0 (->3cba)!)
    localhost.59090 > localhost.krb524: Flags [S], cksum 0xfe30 (incorrect -> 0xf1cb), seq 1148002775, win 65535, options [mss 16344,nop,wscale 6,sackOK,TS val 126793741 ecr 0], length 0
E..<..@.@..............\Dm!..........0....?........
........
11:55:28.366309 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60, bad cksum 0 (->3cba)!)
    localhost.59091 > localhost.krb524: Flags [S], cksum 0xfe30 (incorrect -> 0x30cf), seq 2647033965, win 65535, options [mss 16344,nop,wscale 6,sackOK,TS val 126794009 ecr 0], length 0
E..<..@.@..............\...m.........0....?........
........
11:55:28.366390 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59091 > localhost.krb524: Flags [.], cksum 0xfe28 (incorrect -> 0xf1dd), ack 921507777, win 1276, options [nop,nop,TS val 126794009 ecr 36786677], length 0
E..4..@.@..............\...n6........(.....
.....1Q.
11:55:28.366553 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59091 > localhost.krb524: Flags [F.], cksum 0xfe28 (incorrect -> 0xf1dc), seq 0, ack 1, win 1276, options [nop,nop,TS val 126794009 ecr 36786677], length 0
E..4..@.@..............\...n6........(.....
.....1Q.
11:55:28.366788 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59091 > localhost.krb524: Flags [.], cksum 0xfe28 (incorrect -> 0xf1db), ack 2, win 1276, options [nop,nop,TS val 126794009 ecr 36786677], length 0
E..4..@.@..............\...o6........(.....
.....1Q.
11:55:28.368777 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 60, bad cksum 0 (->3cba)!)
    localhost.59092 > localhost.krb524: Flags [S], cksum 0xfe30 (incorrect -> 0x8f63), seq 224246335, win 65535, options [mss 16344,nop,wscale 6,sackOK,TS val 126794011 ecr 0], length 0
E..<..@.@..............\.].?.........0....?........
........
11:55:28.368854 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59092 > localhost.krb524: Flags [.], cksum 0xfe28 (incorrect -> 0x8c5c), ack 1472971426, win 1276, options [nop,nop,TS val 126794011 ecr 3471146390], length 0
E..4..@.@..............\.].@W........(.....
......}.
11:55:28.369139 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 294, bad cksum 0 (->3bd0)!)
    localhost.59092 > localhost.krb524: Flags [P.], cksum 0xff1a (incorrect -> 0xd3f1), seq 0:242, ack 1, win 1276, options [nop,nop,TS val 126794011 ecr 3471146390], length 242
E..&..@.@..............\.].@W..............
......}.POST /session HTTP/1.1
Accept: application/json
Content-Type: application/json; charset=UTF-8
User-Agent: selenium/3.142.2 (ruby unix)
Content-Length: 394
Accept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3
Host: 127.0.0.1:4444


11:55:28.369189 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 446, bad cksum 0 (->3b38)!)
    localhost.59092 > localhost.krb524: Flags [P.], cksum 0xffb2 (incorrect -> 0x7098), seq 242:636, ack 1, win 1276, options [nop,nop,TS val 126794011 ecr 3471146390], length 394
E.....@.@..............\.].2W..............
......}.{"desiredCapabilities":{"browserName":"firefox","version":"","platform":"ANY","javascriptEnabled":false,"cssSelectorsEnabled":false,"takesScreenshot":false,"nativeEvents":false,"rotatable":false,"marionette":true,"timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}},"capabilities":{"firstMatch":[{"browserName":"firefox","timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}}]}}
11:55:28.370953 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59092 > localhost.krb524: Flags [.], cksum 0xfe28 (incorrect -> 0x883f), ack 415, win 1276, options [nop,nop,TS val 126794013 ecr 3471146391], length 0
E..4..@.@..............\.]..W..@.....(.....
......}.
11:55:28.482644 IP (tos 0x0, ttl 64, id 0, offset 0, flags [DF], proto TCP (6), length 52, bad cksum 0 (->3cc2)!)
    localhost.59092 > localhost.krb524: Flags [F.], cksum 0xfe28 (incorrect -> 0x87ce), seq 636, ack 415, win 1276, options [nop,nop,TS val 126794125 ecr 3471146391], length 0
E..4..@.@..............\.]..W..@.....(.....
......}.
^C
11 packets captured
288 packets received by filter
0 packets dropped by kernel
```

**Comment by PeterMozesMerl** [comment]
I realized I should have run it with the framebuffer display just like Jenkins does. This might be more useful:

```
X.Org X Server 1.18.4
Release Date: 2016-07-19
X Protocol Version 11, Revision 0
Build Operating System: FreeBSD 11.2-RELEASE-p9 amd64
Current Operating System: FreeBSD 11.2-RELEASE-p10 FreeBSD 11.2-RELEASE-p10 #0: Mon May 13 21:20:50 UTC 2019     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC amd64
Build Date: 07 February 2019  09:44:50AM

Current version of pixman: 0.34.0
	Before reporting problems, check http://wiki.x.org
	to make sure that you have the latest version.
Markers: (--) probed, (**) from config file, (==) default setting,
	(++) from command line, (!!) notice, (II) informational,
	(WW) warning, (EE) error, (NI) not implemented, (??) unknown.
(==) Log file: "/var/log/Xorg.0.log", Time: Mon May 20 13:21:09 2019
(++) Using config file: "/usr/local/etc/X11/xorg.conf.jenkins"
2019-05-20 13:21:12 DEBUG Selenium Executing Process ["/usr/local/jenkins/bin/geckodriver", "--binary=/usr/local/bin/firefox", "--port=4444"]
2019-05-20 13:21:12 DEBUG Selenium polling for socket on ["127.0.0.1", 4444]
2019-05-20 13:21:12 INFO Selenium Skipping handshake as we know it is W3C.
2019-05-20 13:21:12 INFO Selenium -> POST session
2019-05-20 13:21:12 INFO Selenium    >>> http://127.0.0.1:4444/session | {"desiredCapabilities":{"browserName":"firefox","marionette":true,"timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}},"capabilities":{"firstMatch":[{"browserName":"firefox","timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}}]}}
2019-05-20 13:21:12 DEBUG Selenium      > {"Accept"=>"application/json", "Content-Type"=>"application/json; charset=UTF-8", "User-Agent"=>"selenium/3.141.0 (ruby unix)", "Content-Length"=>"247"}
1558351272721	mozrunner::runner	INFO	Running command: "/usr/local/bin/firefox" "-marionette" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofile.UYLc3ZKo0Wqg"
1558351272722	geckodriver::marionette	DEBUG	Waiting 60s to connect to browser on 127.0.0.1:51162
1558351273579	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
1558351273580	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
1558351273580	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
1558351273719	Marionette	TRACE	Received observer notification profile-after-change
1558351273766	Marionette	TRACE	Received observer notification command-line-startup
1558351273766	Marionette	TRACE	Received observer notification nsPref:changed
1558351273766	Marionette	DEBUG	Init aborted (running=false, enabled=true, finalUIStartup=false)
1558351273911	Marionette	TRACE	Received observer notification toplevel-window-ready
1558351275099	Marionette	TRACE	Received observer notification sessionstore-windows-restored
1558351275099	Marionette	TRACE	Waiting until startup recorder finished recording startup scripts...
1558351275359	Marionette	TRACE	All scripts recorded.
1558351275359	Marionette	DEBUG	Setting recommended pref apz.content_response_timeout to 60000
1558351275359	Marionette	DEBUG	Setting recommended pref browser.contentblocking.introCount to 99
1558351275359	Marionette	DEBUG	Setting recommended pref browser.download.panel.shown to true
1558351275359	Marionette	DEBUG	Setting recommended pref browser.newtabpage.enabled to false
1558351275360	Marionette	DEBUG	Setting recommended pref browser.pagethumbnails.capturing_disabled to true
1558351275360	Marionette	DEBUG	Setting recommended pref browser.search.update to false
1558351275361	Marionette	DEBUG	Setting recommended pref browser.tabs.disableBackgroundZombification to false
1558351275361	Marionette	DEBUG	Setting recommended pref browser.tabs.warnOnCloseOtherTabs to false
1558351275361	Marionette	DEBUG	Setting recommended pref browser.tabs.warnOnOpen to false
1558351275361	Marionette	DEBUG	Setting recommended pref browser.usedOnWindows10.introURL to
1558351275361	Marionette	DEBUG	Setting recommended pref browser.urlbar.suggest.searches to false
1558351275362	Marionette	DEBUG	Setting recommended pref datareporting.policy.dataSubmissionPolicyAccepted to false
1558351275362	Marionette	DEBUG	Setting recommended pref dom.disable_beforeunload to true
1558351275362	Marionette	DEBUG	Setting recommended pref dom.disable_open_during_load to false
1558351275362	Marionette	DEBUG	Setting recommended pref dom.file.createInChild to true
1558351275363	Marionette	DEBUG	Setting recommended pref dom.max_chrome_script_run_time to 0
1558351275363	Marionette	DEBUG	Setting recommended pref dom.max_script_run_time to 0
1558351275363	Marionette	DEBUG	Setting recommended pref extensions.getAddons.cache.enabled to false
1558351275363	Marionette	DEBUG	Setting recommended pref extensions.webservice.discoverURL to http://%(server)s/dummy/discoveryURL
1558351275363	Marionette	DEBUG	Setting recommended pref network.http.prompt-temp-redirect to false
1558351275364	Marionette	DEBUG	Setting recommended pref network.http.speculative-parallel-limit to 0
1558351275364	Marionette	DEBUG	Setting recommended pref security.fileuri.strict_origin_policy to false
1558351275364	Marionette	DEBUG	Setting recommended pref security.notification_enable_delay to 0
1558351275364	Marionette	DEBUG	Setting recommended pref signon.autofillForms to false
1558351275364	Marionette	DEBUG	Setting recommended pref signon.rememberSignons to false
1558351275364	Marionette	DEBUG	Setting recommended pref toolkit.cosmeticAnimations.enabled to false
1558351275412	Marionette	INFO	Listening on port 51162
1558351275412	Marionette	DEBUG	Remote service is active
1558351275470	geckodriver::marionette	DEBUG	Connection established on 127.0.0.1:51162. Waiting for Marionette handshake
1558351275471	Marionette	DEBUG	Accepted connection 0 from 127.0.0.1:51193
1558351275471	geckodriver::marionette	DEBUG	Connected to Marionette
1558351275472	Marionette	DEBUG	0 -> [0,1,"WebDriver:NewSession",{"browserName":"firefox","timeouts":{}}]
1558351275513	Marionette	TRACE	[4294967297] Frame script loaded
1558351275514	Marionette	TRACE	[4294967297] Frame script registered
1558351275519	Marionette	DEBUG	0 <- [1,1,null,{"sessionId":"0380cf30-daa2-4bec-b944-bb2d78233226","capabilities":{"browserName":"firefox","browserVersion":"66.0. ... ozprofile.UYLc3ZKo0Wqg","moz:shutdownTimeout":60000,"moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true}}]
1558351275522	webdriver::server	DEBUG	<- 200 OK {"value":{"sessionId":"0380cf30-daa2-4bec-b944-bb2d78233226","capabilities":{"acceptInsecureCerts":false,"browserName":"firefox","browserVersion":"66.0.5","moz:accessibilityChecks":false,"moz:geckodriverVersion":"0.24.0","moz:headless":false,"moz:processID":32615,"moz:profile":"/tmp/rust_mozprofile.UYLc3ZKo0Wqg","moz:shutdownTimeout":60000,"moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"freebsd","platformVersion":"11.2-RELEASE-p10","rotatable":false,"setWindowRect":true,"strictFileInteractability":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"unhandledPromptBehavior":"dismiss and notify"}}}
2019-05-20 13:21:15 INFO Selenium <- {"value":{"sessionId":"0380cf30-daa2-4bec-b944-bb2d78233226","capabilities":{"acceptInsecureCerts":false,"browserName":"firefox","browserVersion":"66.0.5","moz:accessibilityChecks":false,"moz:geckodriverVersion":"0.24.0","moz:headless":false,"moz:processID":32615,"moz:profile":"/tmp/rust_mozprofile.UYLc3ZKo0Wqg","moz:shutdownTimeout":60000,"moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"freebsd","platformVersion":"11.2-RELEASE-p10","rotatable":false,"setWindowRect":true,"strictFileInteractability":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"unhandledPromptBehavior":"dismiss and notify"}}}
2019-05-20 13:21:15 INFO Selenium Detected W3C dialect.
2019-05-20 13:21:15 INFO Selenium -> DELETE session/0380cf30-daa2-4bec-b944-bb2d78233226
1558351275524	webdriver::server	DEBUG	-> DELETE /session/0380cf30-daa2-4bec-b944-bb2d78233226
1558351275529	Marionette	DEBUG	0 -> [0,2,"Marionette:Quit",{"flags":["eForceQuit"]}]
1558351275530	Marionette	INFO	Stopped listening on port 51162
1558351275555	Marionette	TRACE	Received observer notification quit-application
1558351275560	Marionette	DEBUG	0 <- [1,2,null,{"cause":"shutdown"}]
1558351275585	webdriver::server	DEBUG	Deleting session
1558351275588	Marionette	DEBUG	0 -> [0,3,"Marionette:Quit",{"flags":["eForceQuit"]}]
1558351275589	Marionette	DEBUG	0 <- [1,3,{"error":"invalid session id","message":"Tried to run command without establishing a connection","stacktrace":"WebDriver ... et@chrome://marionette/content/server.js:235:8\n_onJSONObjectReady/<@chrome://marionette/content/transport.js:493:9\n"},null]
1558351275597	Marionette	DEBUG	Closed connection 0
JavaScript error: resource://gre/modules/Sqlite.jsm, line 834: Error: Connection is not open.
1558351275717	Marionette	TRACE	Received observer notification xpcom-will-shutdown
1558351275717	Marionette	DEBUG	Resetting recommended pref apz.content_response_timeout
1558351275717	Marionette	DEBUG	Resetting recommended pref browser.contentblocking.introCount
1558351275717	Marionette	DEBUG	Resetting recommended pref browser.download.panel.shown
1558351275717	Marionette	DEBUG	Resetting recommended pref browser.newtabpage.enabled
1558351275718	Marionette	DEBUG	Resetting recommended pref browser.pagethumbnails.capturing_disabled
1558351275718	Marionette	DEBUG	Resetting recommended pref browser.search.update
1558351275718	Marionette	DEBUG	Resetting recommended pref browser.tabs.disableBackgroundZombification
1558351275718	Marionette	DEBUG	Resetting recommended pref browser.tabs.warnOnCloseOtherTabs
1558351275718	Marionette	DEBUG	Resetting recommended pref browser.tabs.warnOnOpen
1558351275718	Marionette	DEBUG	Resetting recommended pref browser.usedOnWindows10.introURL
1558351275718	Marionette	DEBUG	Resetting recommended pref browser.urlbar.suggest.searches
1558351275719	Marionette	DEBUG	Resetting recommended pref datareporting.policy.dataSubmissionPolicyAccepted
1558351275719	Marionette	DEBUG	Resetting recommended pref dom.disable_beforeunload
1558351275719	Marionette	DEBUG	Resetting recommended pref dom.disable_open_during_load
1558351275719	Marionette	DEBUG	Resetting recommended pref dom.file.createInChild
1558351275719	Marionette	DEBUG	Resetting recommended pref dom.max_chrome_script_run_time
1558351275719	Marionette	DEBUG	Resetting recommended pref dom.max_script_run_time
1558351275719	Marionette	DEBUG	Resetting recommended pref extensions.getAddons.cache.enabled
1558351275720	Marionette	DEBUG	Resetting recommended pref extensions.webservice.discoverURL
1558351275720	Marionette	DEBUG	Resetting recommended pref network.http.prompt-temp-redirect
1558351275720	Marionette	DEBUG	Resetting recommended pref network.http.speculative-parallel-limit
1558351275720	Marionette	DEBUG	Resetting recommended pref security.fileuri.strict_origin_policy
1558351275720	Marionette	DEBUG	Resetting recommended pref security.notification_enable_delay
1558351275720	Marionette	DEBUG	Resetting recommended pref signon.autofillForms
1558351275720	Marionette	DEBUG	Resetting recommended pref signon.rememberSignons
1558351275720	Marionette	DEBUG	Resetting recommended pref toolkit.cosmeticAnimations.enabled
1558351275721	Marionette	DEBUG	Remote service is inactive
1558351276226	geckodriver::marionette	DEBUG	Browser process stopped: exit code: 0
1558351276230	webdriver::server	DEBUG	<- 200 OK {"value":null}
2019-05-20 13:21:16 INFO Selenium <- {"value":null}
Server terminated successfully (0). Closing log file.
```


```
X.Org X Server 1.18.4
Release Date: 2016-07-19
X Protocol Version 11, Revision 0
Build Operating System: FreeBSD 11.2-RELEASE-p9 amd64
Current Operating System: FreeBSD 11.2-RELEASE-p10 FreeBSD 11.2-RELEASE-p10 #0: Mon May 13 21:20:50 UTC 2019     root@amd64-builder.daemonology.net:/usr/obj/usr/src/sys/GENERIC amd64
Build Date: 07 February 2019  09:44:50AM

Current version of pixman: 0.34.0
	Before reporting problems, check http://wiki.x.org
	to make sure that you have the latest version.
Markers: (--) probed, (**) from config file, (==) default setting,
	(++) from command line, (!!) notice, (II) informational,
	(WW) warning, (EE) error, (NI) not implemented, (??) unknown.
(==) Log file: "/var/log/Xorg.0.log", Time: Mon May 20 13:24:47 2019
(++) Using config file: "/usr/local/etc/X11/xorg.conf.jenkins"
2019-05-20 13:24:50 DEBUG Selenium Executing Process ["/usr/local/jenkins/bin/geckodriver", "--port=4444"]
2019-05-20 13:24:50 DEBUG Selenium polling for socket on ["127.0.0.1", 4444]
2019-05-20 13:24:50 INFO Selenium Skipping handshake as we know it is W3C.
2019-05-20 13:24:50 INFO Selenium -> POST session
2019-05-20 13:24:50 INFO Selenium    >>> http://127.0.0.1:4444/session | {"desiredCapabilities":{"browserName":"firefox","version":"","platform":"ANY","javascriptEnabled":false,"cssSelectorsEnabled":false,"takesScreenshot":false,"nativeEvents":false,"rotatable":false,"marionette":true,"timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}},"capabilities":{"firstMatch":[{"browserName":"firefox","timeouts":{},"moz:firefoxOptions":{"log":{"level":"trace"}}}]}}
2019-05-20 13:24:50 DEBUG Selenium      > {"Accept"=>"application/json", "Content-Type"=>"application/json; charset=UTF-8", "User-Agent"=>"selenium/3.142.2 (ruby unix)", "Content-Length"=>"394"}
1558351490594	webdriver::server	DEBUG	<- 500 Internal Server Error {"value":{"error":"session not created","message":"Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line","stacktrace":""}}
2019-05-20 13:24:50 INFO Selenium <- {"value":{"error":"session not created","message":"Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line","stacktrace":""}}
Traceback (most recent call last):
	31: from bin/rails:4:in `<main>'
	30: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
	29: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
	28: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `block in require'
	27: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	26: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require_with_bootsnap_lfi'
	25: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
	24: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `block in require_with_bootsnap_lfi'
	23: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require'
	22: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/commands.rb:18:in `<main>'
	21: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/command.rb:46:in `invoke'
	20: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/command/base.rb:65:in `perform'
	19: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	18: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	17: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	16: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/railties-5.2.3/lib/rails/commands/runner/runner_command.rb:38:in `perform'
	15: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:59:in `load'
	14: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/bootsnap-1.4.4/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:59:in `load'
	13: from /usr/local/jenkins/workspace/xxx/selenium.rb:6:in `<main>'
	12: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver.rb:88:in `for'
	11: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/common/driver.rb:54:in `for'
	10: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
	 9: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/driver.rb:33:in `new'
	 8: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/firefox/marionette/driver.rb:44:in `initialize'
	 7: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:102:in `create_session'
	 6: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/bridge.rb:167:in `execute'
	 5: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:64:in `call'
	 4: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/default.rb:114:in `request'
	 3: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `create_response'
	 2: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/http/common.rb:88:in `new'
	 1: from /usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:34:in `initialize'
/usr/local/jenkins/.rvm/gems/ruby-2.6.3/gems/selenium-webdriver-3.142.2/lib/selenium/webdriver/remote/response.rb:72:in `assert_ok': Expected browser binary location, but unable to find binary in default location, no 'moz:firefoxOptions.binary' capability provided, and no binary flag set on the command line (Selenium::WebDriver::Error::SessionNotCreatedError)
Server terminated successfully (0). Closing log file.
```

**Comment by PeterMozesMerl** [comment]
Okay, I figured it out. It’s not the POST data. Although the v3.141.0 works from Rails, if I send the same data to geckodriver via curl, the error message is the same as it is for v3.142.2.

It is how geckodriver is executed by Selenium.

```Selenium Executing Process ["/usr/local/jenkins/bin/geckodriver", "--binary=/usr/local/bin/firefox", "--port=4444"]```

```Selenium Executing Process ["/usr/local/jenkins/bin/geckodriver", "--port=4444"]```

**Comment by p0deje** [comment]
Thank you for triaging an issue, this is a regression introduced in 0272df572a3fdab07b88a60e74b4944eb1c61545. Can you please try the following patch (or monkey patch) and see if it resolves an issue:

```diff
diff --git a/rb/lib/selenium/webdriver/common/service.rb b/rb/lib/selenium/webdriver/common/service.rb
index e55d212d4f..d4d8bd5436 100644
--- a/rb/lib/selenium/webdriver/common/service.rb
+++ b/rb/lib/selenium/webdriver/common/service.rb
@@ -41,8 +41,11 @@ module Selenium
           Chrome::Service.new(*args)
         end

-        def firefox(*args)
-          Firefox::Service.new(*args)
+        def firefox(**opts)
+          args = opts.delete(:args)
+          opts[:args] = ["--binary=#{Firefox::Binary.path}"]
+          opts[:args] += args if args
+          Firefox::Service.new(**opts)
         end

         def ie(*args)
```

**Comment by PeterMozesMerl** [comment]
You’re welcome. Yes, it does.

Thank you for the quick fix and the work on Selenium in general.

**Comment by p0deje** [comment]
Thank you for the bug report and triaging!

I've release 3.142.3 with the fix, so you can now update. On a side note, it seems like this might also be valuable to raise an issue for GeckoDriver so that it also searches in `/usr/local/bin` for FIrefox binary by default.
